scripts/build_images/build.py

```
@@ -0,0 +1,244 @@
+            f"localhost:5000/{image}@{digest}",
+            f"localhost:5000/{image}:{tag}",
+        ])
+        print_and_run_command([
+            "docker",
+            "push",
+            f"localhost:5000/{image}:{tag}",
+        ])
+
+    # Write metadata to the metadata directory
+    if metadata_dir:
+        metadata = json.loads(metadata)
+        build_ref = metadata["buildx.build.ref"]
+
+        # If pushing is enabled, the image name will be a comma-separated list of image names.
+        # We'll replace it with just the plain image name.
+        metadata["image.name"] = image
+
+        # We need a unique name for the metadata file. We'll use the part of the
+        # image name after the last slash, and a hash of the build ref.
+        name_without_scope = image.split("/")[-1]
+        hashed_build_ref = hashlib.sha256(build_ref.encode()).hexdigest()
+        metadata_filename = f"{name_without_scope}_{hashed_build_ref}.json"
+        with open(os.path.join(metadata_dir, metadata_filename), "w") as f:
+            json.dump(metadata, f, indent=2)
+
+
+def build_images(
+    images: list[str],
+    *,
+    tag: str,
+    platform: str,
+    builder: str,
+    should_push: bool = False,
+    only_changed: bool = False,
+    metadata_dir: str | None = None,
+) -> None:
+    built_images: set[str] = set()
+
+    for image in images:
+        base_image = BASE_IMAGE_MAPPING.get(image)
+        base_image_built = base_image in built_images
+        image_path = get_image_path(image)
+        was_modified = not only_changed or check_path_modified(image_path)
+
+        if not was_modified and not base_image_built:
+            print(f"Skipping {image} because it hasn't changed.")
+            continue
+
+        build_image(
+            image,
+            tag=tag,
+            platform=platform,
+            builder=builder,
+            should_push=should_push,
+            use_local_base_image=bool(base_image and base_image_built),
```

This is wrong I think: If `images` looks like `[prairielearn, plbase]`, it won't build plbase first. It will just use the remote `plbase`, and build `plbase` afterward.

---

`images` must not look like that. Read down a few more lines:

```py
    # Note that base images must come first in the list so that they're built first.
```

The alternative is of course to topologically sort things based on `BASE_IMAGE_MAPPING` but I _really_ did not want to do that 🙃 it seems like a simple enough rule that we just have to list base images first in the list.

---

Can we throw an error if you don't do that? This feels like a very scary rough edge that one could forget when running locally.

```
if base_image in images and not base_image_built
...
```

---

I agree topologically sorting is annoying.

---

Good point, we can easily detect this case and error. Done in 0fb4a91.

---

Actually, that doesn't work; what if the base image hasn't changed and thus hasn't been built?

---

I removed the check for now in 6f55276 so I can keep working on your other comments.

---

```
if base_image in images and not base_image_built and check_path_modified(get_image_path(base_image))
```
?

---

I went a slightly different direction in a6c5926. Given a list of images, we can just check upfront if it's in the right order.